### ISSUE TITLE
Saves fail with "Error, expected `_id` to be unique" 

### ISSUE THREAD

**Issue by BorntraegerMarc** [blakehaswell/mongoose-unique-validator#88]
I have this code:
```js
const localUser: UserDocument = await this.findById(user.id);
localUser.set(updatedData);
return await localUser.save();
```

this code returns the following error:
```
{"errors":{"_id":{"message":"Error, expected `_id` to be unique. Value: `5881d5a92c9a170011275318`","name":"ValidatorError","properties":{"message":"Error, expected `{PATH}` to be unique. Value: `{VALUE}`","type":"unique","path":"_id","value":"5881d5a92c9a170011275318","reason":{}},"kind":"unique","path":"_id","value":"5881d5a92c9a170011275318","reason":{},"$isValidatorError":true},"displayName":{"message":"Error, expected `displayName` to be unique. Value: `drmuster`","name":"ValidatorError","properties":{"message":"Error, expected `{PATH}` to be unique. Value: `{VALUE}`","type":"unique","path":"displayName","value":"drmuster","reason":{}},"kind":"unique","path":"displayName","value":"drmuster","reason":{},"$isValidatorError":true}},"_message":"users validation failed","message":"users validation failed: _id: Error, expected `_id` to be unique. Value: `5881d5a92c9a170011275318`, displayName: Error, expected `displayName` to be unique. Value: `drmuster`","name":"ValidationError"}
```
in version 2.0.2
This used to work with version 2.0.1...

**Comment by viveleroi** [comment]
If it worked in 2.0.1, but doesn't in 2.0.2, the only conclusion is that somehow it's affected by the [countDocuments change](https://github.com/blakehaswell/mongoose-unique-validator/commit/785fa2f2bebf9778b59f4f79115daff6dd340cdf), since that was the only change included. Please be sure you're using a recent version of mongoose

**Comment by BorntraegerMarc** [comment]
I'm using mongoose version 5.0.2

**Comment by gsandorx** [comment]
I'm having the exact same problem. In my case, the error comes up when saving a new document. 

I'm using mongoose 5.3.2. I switched back to mongoose-unique-validator 2.0.1 and the issue is gone, so it's definitively something in 2.0.2

**Comment by viveleroi** [comment]
Do you have the issue if you use `estimatedDocumentCount` instead of `countDocuments`? 

They both are recommended alternates for `count`, but sounds like it's not a 1:1 change.

**Comment by gsandorx** [comment]
Yes, I tried both. I even read that there are operators not supported by `countDocuments` such as `where` and the recommended alternative was `expr`. I tried `model.expr({ $and: conditions }).countDocuments....` (line 102) and same thing.

**Comment by viveleroi** [comment]
Is this still happening with the latest mongoose and mongodb? I can't reproduce it.

It looks like it happens for both new records and existing records but somehow isn't showing up for anything in the current unit tests.

**Comment by gsandorx** [comment]
Yes, same error... I'm using:

"mongoose": "^5.5.5",
"mongoose-unique-validator": "^2.0.3",

**Comment by iBaff** [comment]
I also have the same problem with `save()`:

"mongoose": "^5.7.1",
"mongoose-unique-validator": "^2.0.3",

on *MongoDB Atlas - M0 Sandbox*

> ValidatorError: Error, expected `_id` to be unique.

and also for other fields with `unique: true`

> ValidatorError: Error, expected `...` to be unique.

**Comment by jgall** [comment]
Are there any known ways to fix this issue? I'm running into it as well.

**Comment by Ding-Fan** [comment]
> I also have the same problem with `save()`:
> ......
> But it would be super handy if the validator could also do it's work on the free/sandbox tier.

Same error here. Drop my database and create a new one solved my problem. 👀

**Comment by j-tesla** [comment]
@viveleroi I'm facing the same issue and [here](https://github.com/j-tesla/blog-list/pull/58/checks?check_run_id=3845735523) are some unit tests which failed.

**Comment by Suryapratap-R** [comment]
is there any way to do this with version 3, facing this issue

**Comment by saitama1899** [comment]
i'm facing this issue with 
    "mongoose": "6.1.7",
    "mongoose-unique-validator": "3.0.0"
i'm using mongo db cloud

**Comment by BlazingTide** [comment]
Facing the same issue, just began happening. Moving back to ^2.0.3 to see if it fixes it.

**Comment by rubek-joshi** [comment]
I can also confirm that downgrading the package to version 2.0.3 worked.

**Comment by AdamKniec** [comment]
I had the same issue. Could not use the 3.0.0 validator and `.save` method from mongo. Downgrading the package worked

**Comment by Dirk-** [comment]
I started to get this issue when MongoDB Atlas upgraded my cluster to MongoDB 5.0.

It used to work with mongoose 6.0.12 and mongoose-unique-validator 3.0.0 on pre-MongoDB 5.0, now it doesn't work any more. mongoose 6.2.0 does not help, and going back to mongoose-unique-validator 2.0.3 is not possible because it needs mongoose@"^5.2.1".

@viveleroi I can share my schema and controller if you want.

**Comment by harrisonellis16** [comment]
SOLVED WITH: 

await model.save({
    validateModifiedOnly: true,
  })

This save option solved it for me without having to change version, because it makes it validate only the updated fields meaning it won't recognize ID as being a duplicate.

Currently using 
"mongoose": "6.2.0",
"mongoose-unique-validator": "3.0.0"

**Comment by K-G-J** [comment]
> 

Can you please post your full code for the await function you used.  I am having the same problem

**Comment by AmirHo3einJS** [comment]
> SOLVED WITH:
> 
> ```
> await model.save({
>     validateModifiedOnly: true,
>   })
> ```
> 
> This save option solved it for me without having to change version, because it makes it validate only the updated fields meaning it won't recognize ID as being a duplicate.
> 
> Currently using "mongoose": "6.2.0", "mongoose-unique-validator": "3.0.0"

This solution works,
using: {
     "mongoose": "^6.2.7",
     "mongoose-unique-validator": "^3.0.0",
}

**Comment by FilipMl** [comment]
validateModifiedOnly: true solved for me. Thanks!

**Comment by ArnieGA** [comment]
> SOLVED WITH:
> 
> ```
> await model.save({
>     validateModifiedOnly: true,
>   })
> ```
> 
> This save option solved it for me without having to change version, because it makes it validate only the updated fields meaning it won't recognize ID as being a duplicate.
> 
> Currently using "mongoose": "6.2.0", "mongoose-unique-validator": "3.0.0"

Using `mongoose@6.3.4 | mongoose-unique-validator@3.0.0`, the solution described works.
**However**, respectfully, this should not be considered as the official solution because it is a limiting solution.
I'd be limited to always having to include that object on every save, and limited to validating modified-only paths.

**Comment by viveleroi** [comment]
This should be fixed in #109 but it's difficult for me to tell if there are multiple scenarios that lead to this issue. I'll be releasing an update soon so please report any unresolved issues.

**Comment by ArnieGA** [comment]
> This should be fixed in #109 but it's difficult for me to tell if there are multiple scenarios that lead to this issue. I'll be releasing an update soon so please report any unresolved issues.

Will do.  Please let us know when the updates are live.

**Comment by viveleroi** [comment]
v3.1.0 is out now. 🤞
